edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs

```diff
@@ -0,0 +1,747 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Controllers
+{
+    using System;
+    using System.Net;
+    using System.Net.Http;
+    using System.Text;
+    using System.Threading.Tasks;
+    using System.Web;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class RegistryController : Controller
+    {
+        const string ContentType = "application/json; charset=utf-8";
+
+        readonly IRegistryApiClient apiClient;
+        readonly Task<IHttpRequestAuthenticator> authenticatorGetter;
+        readonly Task<IEdgeHub> edgeHubGetter;
+
+        public RegistryController(
+            IRegistryApiClient apiClient,
+            Task<IEdgeHub> edgeHubGetter,
+            Task<IHttpRequestAuthenticator> authenticatorGetter)
+        {
+            this.apiClient = Preconditions.CheckNotNull(apiClient, nameof(apiClient));
+            this.authenticatorGetter = Preconditions.CheckNotNull(authenticatorGetter, nameof(authenticatorGetter));
+            this.edgeHubGetter = Preconditions.CheckNotNull(edgeHubGetter, nameof(edgeHubGetter));
+        }
+
+        [HttpPut]
+        [Route("devices/{deviceId}/modules/{moduleId}")]
+        public async Task CreateOrUpdateModuleAsync(
+            [FromRoute] string deviceId,
+            [FromRoute] string moduleId,
+            [FromBody] Module module)
+        {
+            try
+            {
+                Events.ReceivedRequest(nameof(this.CreateOrUpdateModuleAsync), deviceId, moduleId);
+
+                try
+                {
+                    deviceId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId)));
+                    moduleId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(moduleId, nameof(moduleId)));
+                    Preconditions.CheckNotNull(module, nameof(module));
+
+                    if (!string.Equals(deviceId, module.DeviceId) || !string.Equals(moduleId, module.Id))
+                    {
+                        throw new ApplicationException("Device Id or module Id doesn't match between request URI and body.");
+                    }
```

Comment:
Interesting - why are the deviceId and moduleId specified twice in the request? Do we know?

Reply:
as mentioned, this is existing registry api method.